### PR TITLE
fasthttptrace: Add trace context propagation support for libraries built on fasthttp

### PR DESCRIPTION
**Pull request by mtoffl01**
### What does this PR do?
- Add foundational support for tracing libraries built on fasthttp, including a fasthttp implementation of TextMapCarrier and a fasthttp implementation of StartSpanFromContext
- Move the contextKey type and ActiveSpanKey into ddtrace `internal` package and make ActiveSpanKey accessible to other packages that may need it (such as fasthttp contrib packages)

### Motivation
In adding a contrib for [gearbox](https://pkg.go.dev/github.com/gogearbox/gearbox) (PR: https://github.com/DataDog/dd-trace-go/pull/2118), we found that the existing functionality in ddtrace relied heavily on a net/http implementation. We needed new functionality to update the fasthttp request context and to propagate trace context via fasthttp request headers. 

### Reviewer's Checklist

- [ ] Changed code has unit tests for its functionality at or near 100% coverage.
- [ ] There is a benchmark for any new code, or changes to existing code.
- [ ] If this interacts with the agent in a new way, a system test has been added.

For Datadog employees:

- [ ] If this PR touches code that handles credentials of any kind, such as Datadog API keys, I've requested a review from `@DataDog/security-design-and-guidance`.
- [X] This PR doesn't touch any of that.

Unsure? Have a question? Request a review!

### CODE REVIEW

File: contrib/internal/fasthttptrace/fasthttpheaderscarrier.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+		}
+	}
+	return nil
+}
+
+// Set adds the given value to request header for key. Key will be lowercased to match
+// the metadata implementation.
+func (f *FastHTTPHeadersCarrier) Set(key, val string) {
+	// f.ReqHeader.Set(k, val)
+	// MOTFF: "Set" overwrites any value at `k`. "Add" appends it. Just confirming we want to append, not overwrite
```
Comment (mtoffl01) [inline contrib/internal/fasthttptrace/fasthttpheaderscarrier.go:43]: See discussion here: https://github.com/DataDog/dd-trace-go/pull/2118#discussion_r1321562999
TLDR; 
The definition for TextMapCarrier's Set method is nested and very vague: "Set implements TextMapWriter."
...The definition for TextMapWriter's Set method states, "Set sets the given key/value pair."

grpc's [mdcarrier](https://github.com/DataDog/dd-trace-go/blob/main/contrib/google.golang.org/internal/grpcutil/mdcarrier.go#L32-L37), which fulfills TextMapCarrier, appends rather than overwrites: https://github.com/DataDog/dd-trace-go/blob/main/contrib/google.golang.org/internal/grpcutil/mdcarrier.go#L32-L37

It seems that we'd just want to overwrite data, but I want to make sure this gets properly reviewed.
**Do we want to append or set/overwrite data?**

File: contrib/internal/fasthttptrace/fasthttpheaderscarrier_test.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+	fcc := &FastHTTPHeadersCarrier{
+		ReqHeader: &fasthttp.RequestHeader{},
+	}
+	err := tracer.Inject(pspan.Context(), fcc)
+	if err != nil {
+		t.Fatal("Creating new request with context failed")
+	}
+	if err != nil {
+		t.Fatal("Request failed")
+	}
```
Comment (rarguelloF) [inline contrib/internal/fasthttptrace/fasthttpheaderscarrier_test.go:98]:
```suggestion
	require.NoError(t, err)
```

File: contrib/internal/fasthttptrace/fasthttpheaderscarrier_test.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	sctx, err := tracer.Extract(fcc)
+	if err != nil {
+		t.Fatal("Trace extraction failed")
+	}
```
Comment (rarguelloF) [inline contrib/internal/fasthttptrace/fasthttpheaderscarrier_test.go:102]:
```suggestion
	require.NoError(t, err)
```

File: contrib/internal/fasthttptrace/fasthttptrace_test.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+func TestStartSpanFromFastHTTPContext(t *testing.T) {
+	assert := assert.New(t)
+	mt := mocktracer.Start()
+	defer mt.Stop()
+	fctx := &fasthttp.RequestCtx{}
+	activeSpan := StartSpanFromFastHTTPContext(fctx, "myOp")
+	keySpan := fctx.UserValue(internal.ActiveSpanKey)
+	assert.Equal(activeSpan, keySpan)
+
```
Comment (rarguelloF) [inline contrib/internal/fasthttptrace/fasthttptrace_test.go:25]:
```suggestion
```

File: ddtrace/tracer/context.go [hunk 5]
```diff
@@ -9,30 +9,27 @@ import (
 	"context"
 
 	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
-	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/internal"
+	tracerInternal "gopkg.in/DataDog/dd-trace-go.v1/ddtrace/internal"
```
Comment (rarguelloF) [inline ddtrace/tracer/context.go:12]:
```suggestion
	traceinternal "gopkg.in/DataDog/dd-trace-go.v1/ddtrace/internal"
```

let's use the same alias as we are using in other files (https://github.com/DataDog/dd-trace-go/blob/main/ddtrace/tracer/transport.go#L22)

File: ddtrace/tracer/context_test.go [hunk 6]
```diff
@@ -12,15 +12,16 @@ import (
 	"testing"
 
 	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
-	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/internal"
+	tracerInternal "gopkg.in/DataDog/dd-trace-go.v1/ddtrace/internal"
```
Comment (rarguelloF) [inline ddtrace/tracer/context_test.go:15]:
```suggestion
	traceinternal "gopkg.in/DataDog/dd-trace-go.v1/ddtrace/internal"
```

File: contrib/internal/fasthttptrace/fasthttpheaderscarrier_test.go [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+	fcc := &FastHTTPHeadersCarrier{
+		ReqHeader: h,
+	}
+	err := fcc.ForeachKey(func(k, v string) error {
+		delete(headers, k)
+		return nil
+	})
+
+	assert.Nil(err)
```
Comment (rarguelloF) [inline contrib/internal/fasthttptrace/fasthttpheaderscarrier_test.go:80]:
```suggestion
	assert.NoError(err)
```

File: contrib/internal/fasthttptrace/fasthttpheaderscarrier_test.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+		}
+	}
+	fcc := &FastHTTPHeadersCarrier{
+		ReqHeader: h,
+	}
+	err := fcc.ForeachKey(func(k, v string) error {
+		delete(headers, k)
+		return nil
+	})
+
```
Comment (rarguelloF) [inline contrib/internal/fasthttptrace/fasthttpheaderscarrier_test.go:79]:
```suggestion
```

File: contrib/internal/fasthttptrace/fasthttpheaderscarrier_test.go [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+		assert.Equal("v2", string(vals[1]))
+	})
+	t.Run("multi-key", func(t *testing.T) {
+		// // add a second item
+		fcc.Set("k2", "v21")
+		// assert.Len(fcc.ReqHeader.PeekKeys(), 2) This returns 3 values: k1, k1 and k2...
+		assert.Len(fcc.ReqHeader.PeekAll("k2"), 1)
+		assert.Equal("v21", string(fcc.ReqHeader.Peek("k2")))
+	})
+	// FastHTTP
```
Comment (rarguelloF) [inline contrib/internal/fasthttptrace/fasthttpheaderscarrier_test.go:45]:
```suggestion
```

File: contrib/internal/fasthttptrace/fasthttpheaderscarrier_test.go [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+		fcc.Set("k1", "v2")
+		vals := fcc.ReqHeader.PeekAll("k1")
+		assert.Len(vals, 2)
+		assert.Equal("v1", string(vals[0]))
+		assert.Equal("v2", string(vals[1]))
+	})
+	t.Run("multi-key", func(t *testing.T) {
+		// // add a second item
+		fcc.Set("k2", "v21")
+		// assert.Len(fcc.ReqHeader.PeekKeys(), 2) This returns 3 values: k1, k1 and k2...
```
Comment (rarguelloF) [inline contrib/internal/fasthttptrace/fasthttpheaderscarrier_test.go:41]:
```suggestion
```

**Review by rarguelloF — CHANGES_REQUESTED** [review]
LGTM overall! left a few nit comments

File: contrib/internal/fasthttptrace/fasthttptrace.go [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+	"gopkg.in/DataDog/dd-trace-go.v1/internal"
+
+	"github.com/valyala/fasthttp"
+)
+
+// StartSpanFromFastHTTPContext returns a new span with the given operation name and options.
+// If a span is found in the `fctx`, it will be used as the parent of the resulting span.
+// The resulting span is then set on the `fctx`.
+// If the ChildOf option is passed, it will only be used as the parent if there is no span found in `fctx`.
+func StartSpanFromFastHTTPContext(fctx *fasthttp.RequestCtx, operationName string, opts ...tracer.StartSpanOption) tracer.Span {
```
Comment (ahmed-mez) [inline contrib/internal/fasthttptrace/fasthttptrace.go:19]:
```suggestion
func StartSpanFromContext(fctx *fasthttp.RequestCtx, operationName string, opts ...tracer.StartSpanOption) tracer.Span {
```
To avoid redundancy since the package name is `fasthttptrace`, this will called `fasthttptrace.StartSpanFromContext`.

File: contrib/google.golang.org/grpc/stats_client.go [hunk 12]
```diff
@@ -11,6 +11,7 @@ import (
 
 	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
 	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+
```
Comment (ahmed-mez) [inline contrib/google.golang.org/grpc/stats_client.go:14]: Since we're not changing any code in this file, it doesn't make too much sense to have this extra line as part of this PR. Feel free to do it in a separate PR though!

**Review by ahmed-mez — COMMENTED** [review]
I believe this is the right approach, thanks for working on it!

Added a few minor suggestions but overall the code looks right, let's also rename the PR!

File: contrib/google.golang.org/grpc/stats_client.go (hunk 12 above)
Comment (mtoffl01) [reply contrib/google.golang.org/grpc/stats_client.go:14]: No idea how this happened

File: contrib/internal/fasthttptrace/fasthttpheaderscarrier.go [hunk 13]
```diff
@@ -0,0 +1,45 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
```
[the remark this answers is not in the document]
Comment (mtoffl01) [reply contrib/internal/fasthttptrace/fasthttpheaderscarrier.go:1]: What does this mean?

File: contrib/internal/fasthttptrace/fasthttpheaderscarrier.go [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+type FastHTTPHeadersCarrier struct {
+	ReqHeader *fasthttp.RequestHeader
+}
+
+var _ tracer.TextMapWriter = (*FastHTTPHeadersCarrier)(nil)
+var _ tracer.TextMapReader = (*FastHTTPHeadersCarrier)(nil)
+
+// ForeachKey iterates over fasthttp request header keys and values
+func (f *FastHTTPHeadersCarrier) ForeachKey(handler func(key, val string) error) error {
+	fmt.Println("IN FOR EACH KEY")
```
Comment (ahmed-mez) [inline contrib/internal/fasthttptrace/fasthttpheaderscarrier.go:28]:
```suggestion
```

File: contrib/internal/fasthttptrace/fasthttpheaderscarrier.go [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+		}
+	}
+	return nil
+}
+
+// Set adds the given value to request header for key. Key will be lowercased to match
+// the metadata implementation.
+func (f *FastHTTPHeadersCarrier) Set(key, val string) {
+	f.ReqHeader.Set(key, val)
+	fmt.Printf("req header: %v, value %v\n", key, string(f.ReqHeader.Peek(key)))
```
Comment (ahmed-mez) [inline contrib/internal/fasthttptrace/fasthttpheaderscarrier.go:44]:
```suggestion
```

**Review by ahmed-mez — APPROVED** [review]
LGTM, feel free to merge without a second approval from me once the linter and remaining comments/issues are resolved. @rarguelloF please take a look as well. Thanks!

File: contrib/internal/fasthttptrace/fasthttpheaderscarrier.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (rarguelloF) [reply contrib/internal/fasthttptrace/fasthttpheaderscarrier.go:1]: looks like an old comment, you can ignore 👍

File: contrib/internal/fasthttptrace/fasthttptrace.go [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+package fasthttptrace
+
+import (
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal"
+
+	"github.com/valyala/fasthttp"
+)
+
+// StartSpanFromFastHTTPContext returns a new span with the given operation name and options.
```
Comment (rarguelloF) [inline contrib/internal/fasthttptrace/fasthttptrace.go:15]:
```suggestion
// StartSpanFromContext returns a new span with the given operation name and options.
```

File: contrib/internal/fasthttptrace/fasthttptrace_test.go [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+	"testing"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/mocktracer"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/valyala/fasthttp"
+)
+
+func TestStartSpanFromFastHTTPContext(t *testing.T) {
```
Comment (rarguelloF) [inline contrib/internal/fasthttptrace/fasthttptrace_test.go:18]:
```suggestion
func TestStartSpanFromContext(t *testing.T) {
```

File: contrib/internal/fasthttptrace/fasthttptrace.go (hunk 16 above)
Comment (rarguelloF) [reply contrib/internal/fasthttptrace/fasthttptrace.go:15]: Maybe we could also clarify in this comment that this does the same as `tracer.StartSpanFromContext` but using a fasthttp context.

**Review by rarguelloF — APPROVED** [review]
LGTM! (I just left some comments for missing renames for the StartSpanFromFastHTTPContext -> StartSpanFromContext refactor)
